### ISSUE TITLE
Conditional includes `__cplusplus`

### ISSUE THREAD

**Issue by mnil** [martis42/depend_on_what_you_use#270]
I am trying to uplift to latest `DWYU`, 0.4.0 but hit a blocker with some conditional includes and can't find any docs on whether that is supposed to work or not. From what I can tell do the following example work with 5addf9b13c9e71663d52d334ba09c29b3c2de698 but not with 0.4.0. 

`foo/BUILD.bazel`
```python
cc_library(
    name = "foo",
    hdrs = ["foo.h"],
    deps = [":bar"],
)

cc_library(
    name = "bar",
    hdrs = ["bar.h"],
)
``` 

`foo/foo.h`
```cpp
#ifdef __cplusplus
#include "foo/bar.h"
#endif  // __cplusplus

#define Pi (3.14159)
```

`foo/bar.h`
```cpp
#define e (2.71828)
```

**Comment by martis42** [comment]
Thanks for this error report. I can reproduce this issue.

As far as I can see `__cplusplus` is a special thing which typically is not set from the outside. By outside I mean via `--copt`, the Bazel CC toolchain and similar things. In other words, `__cplusplus` is not a define which is part of the command line arguments forwarded to the compiler to compile your code. This is a problem for DWYU, as we use the defines from the compiler command as input for a customized preprocessing step done with [pcpp](https://github.com/ned14/pcpp). Thus, DWYU treats `__cplusplus` always as undefined. It seems the standard preprocessor and linker set this value internally during their processing.

Your example worked in the past, where we simply ignored preprocessor defines and analyzed a superset of all include statements in a file. While it works in your concrete example it is wrong in most cases, and thus from release `0.0.8` on we use a preprocessor to handle most cases.

I am not yet sure how exactly to resolve your case. Right now I don't see a prefect and general solution. Most likely I will think of some heuristic which one can activate via an opt-in. Meaning, you can expect eventually a fix which allows you to use recent DWYU versions.

**Comment by mnil** [comment]
> Your example worked in the past, where we simply ignored preprocessor defines and analyzed a superset of all include statements in a file.

Does that mean that from 0.0.8 DWYU is sensitive to which toolchain is defined when running DWYU?

**Comment by martis42** [comment]
Actually DWYU was always sensitive to the active toolchain. More precisely, running DWYU is sensitive to the active configuration, of which the toolchain is a part. This is the same for all Bazel aspects. After all, they are executed as part of `bazel build`.
However, before version 0.0.8 DWYU ignored many parts of the configuration.

Being aware of the configuration is important when select statements control the dependency list. Imagine a target depending on different dependencies depending on whether you are on Linux or Windows and also including different headers depending on the active platform. The DWYU aspect sees only the selected dependencies. If we would not know which includes are relevant, DWYU would always report failures for such targets.

**Comment by martis42** [comment]
Unfortunately I am currently quite busy. I plan to fix this and deploy 0.5.0 in September.

**Comment by martis42** [comment]
@mnil 
[Release 0.5.0](https://github.com/martis42/depend_on_what_you_use/releases/tag/0.5.0) contains an experimental feature which should help you. Please have a look.
